helix-rest/src/main/java/org/apache/helix/rest/server/resources/zookeeper/MetadataStoreDirectoryAccessor.java

```diff
@@ -0,0 +1,163 @@
+
+  /**
+   * Gets all metadata store realms in a namespace with the endpoint.
+   *
+   * @return Json representation of all realms.
+   */
+  @GET
+  @Path("/metadata-store-realms")
+  public Response getAllMetadataStoreRealms() {
+    Map<String, Collection<String>> responseMap = new HashMap<>(1);
+    try {
+      Collection<String> realms =
+          getMetadataStoreDirectory().getAllMetadataStoreRealms(getHelixNamespace().getName());
+      responseMap.put(
+          MetadataStoreDirectoryConstants.METADATA_STORE_REALMS_NAME, realms);
+    } catch (NoSuchElementException ex) {
+      return notFound(ex.getMessage());
+    }
+
+    return JSONRepresentation(responseMap);
+  }
+
+  /**
+   * Gets sharding keys mapped at path "HTTP GET /sharding-keys" which returns all sharding keys in
+   * a namespace, or path "HTTP GET /sharding-keys?realm={realmName}" which returns sharding keys in
+   * a realm.
+   *
+   * @param realm Query param in endpoint path
+   * @return Json representation of a map: shardingKeys -> collection of sharding keys.
+   */
+  @GET
+  @Path("/sharding-keys")
+  public Response getShardingKeys(@QueryParam("realm") String realm) {
+    Map<String, Object> responseMap;
+    Collection<String> shardingKeys;
+    try {
+      // If realm is not set in query param, the endpoint is: "/sharding-keys"
+      // to get all sharding keys in a namespace.
+      if (realm == null) {
+        shardingKeys =
+            getMetadataStoreDirectory().getAllShardingKeys(getHelixNamespace().getName());
+        // To avoid allocating unnecessary resource, limit the map's capacity only for
+        // SHARDING_KEYS_NAME.
+        responseMap = new HashMap<>(1);
+      } else {
+        // For endpoint: "/sharding-keys?realm={realmName}"
+        shardingKeys = getMetadataStoreDirectory()
+            .getAllShardingKeysInRealm(getHelixNamespace().getName(), realm);
+        // To avoid allocating unnecessary resource, limit the map's capacity only for
+        // SHARDING_KEYS_NAME and "metadataStoreRealm".
+        responseMap = new HashMap<>(2);
+        responseMap.put("metadataStoreRealm", realm);
+      }
+    } catch (NoSuchElementException ex) {
+      return notFound(ex.getMessage());
+    }
+
+    responseMap.put(MetadataStoreDirectoryConstants.SHARDING_KEYS_NAME, shardingKeys);
+
+    return JSONRepresentation(responseMap);
+  }
+
+  private MetadataStoreDirectory getMetadataStoreDirectory() {
+    if (_metadataStoreDirectory == null) {
+      synchronized (this) {
+        if (_metadataStoreDirectory == null) {
+          Map<String, String> routingZkAddressMap = ImmutableMap
+              .of(getHelixNamespace().getName(), getHelixNamespace().getMetadataStoreAddress());
+          try {
+            _metadataStoreDirectory = new ZkMetadataStoreDirectory(routingZkAddressMap);
+          } catch (InvalidRoutingDataException ex) {
+            // In this case, the InvalidRoutingDataException should not happen because routing
```

Comment:
how so? what if for some reason routingZkAddressMap gives some weird value like "hellotest"?

Reply:
The reason I put this comment is because below code only checks  routingZkAddressMap is null or empty, but not weird value like "hellotest". So the InvalidRoutingDataException here will not happen...
```
  public ZkMetadataStoreDirectory(Map<String, String> routingZkAddressMap)
      throws InvalidRoutingDataException {
    if (routingZkAddressMap == null || routingZkAddressMap.isEmpty()) {
      throw new InvalidRoutingDataException("Routing ZK Addresses given are invalid!");
    }
```